### ISSUE TITLE
SnarkyJS integration error

### ISSUE THREAD

**Issue by directcuteo** [o1-labs/snarkyjs#910]
I want to include the capabilities of SnarkyJS in my web application. The only code I added that uses snarkyjs is the following one.

```
import { SmartContract } from 'snarkyjs';

class Square extends SmartContract {
}
```

This resulted in the error below. The error is coming from a snarkyjs file.
Have I made a mistake?

![image](https://github.com/o1-labs/snarkyjs/assets/37619567/19067af2-ce6a-490e-bbea-065f4da68943)

![image](https://github.com/o1-labs/snarkyjs/assets/37619567/e9bd22dc-911a-4535-807d-efb854aee8f4)

**Comment by directcuteo** [comment]
I am using it in an Angular v15 app which is built using Webpack v5.75.0

**Comment by mitschabaude** [comment]
@ymekuria do you have an idea what could happen here?

**Comment by directcuteo** [comment]
@mitschabaude Do you have any updates regarding this issue? Thanks!

**Comment by mitschabaude** [comment]
@directcuteo I haven't looked into it. It seems that Angular / webpack is trying to compile snarkyjs' web bundle and is doing something wrong. Since snarkyjs' bundle should work as is, without transpilation, maybe there's some option that you can give to webpack to ignore snarkyjs when bundling the code, i.e. leave it untouched?

**Comment by mitschabaude** [comment]
sure! plain html _ js example:

* take `index.js` which is shipped in the npm package at `/dist/web/index.js`
* make sure `index.js` is located next to `index.html`
* serve the following html:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>hello-snarkyjs</title>
    <script type="importmap">
      { "imports": { "snarkyjs": "./index.js" } }
    </script>    
    <script type="module">
      import { Field } from "snarkyjs";
      console.log(Field(-1));
    </script>
  </head>
  <body>
    <div>Check out the console (F12)</div>
  </body>
</html>
```

**Comment by mitschabaude** [comment]
So the point is, it's better if your web framework doesn't try to transform the snarkyjs build output (bundling should be fine; it should know where to resolve it from the package.json "exports" field)

**Comment by directcuteo** [comment]
Your example works, I appreciate!

My opinion:
The thing is that because I import `SmartContract` in an Angular component, Angular will include `o1js` in its bundling/tree-shaking thing. The problem is that this changes the names of functions and variables. By default this should not be an issue. The issue is that `wasm` is imported later and it will call a function from javascript which (and here is the surprise) doesn't exist anymore because of the tree-shaking.
Following this discussion [https://github.com/evanw/esbuild/issues/458](https://github.com/evanw/esbuild/issues/458) if I get it right, there should be some flag added in `o1js` that will not allow the tree-shaking because it has side-effects (`wasm` will fail to call the functions).
What are your thoughts?

**Comment by mitschabaude** [comment]
@directcuteo wow, thanks for investigating! I'll look into this

**Comment by directcuteo** [comment]
Small upd: building `o1js` locally and using it in the Angular app without the minifier resulted that the undefined function is `_mainWorker`

![image](https://github.com/o1-labs/o1js/assets/37619567/b640f3e0-613f-4886-b3db-57455335fdd9)

**Comment by mitschabaude** [comment]
hm that's weird.. this is a transformed version of the code here, which has normal function code inside `mainWorker()`:
https://github.com/o1-labs/o1js-bindings/blob/66061bfd2349c16414330410881fcf6e6473ed6d/js/web/web-backend.js#L117-L124

**Comment by directcuteo** [comment]
That transformed version is coming from a blob file imported via http request.

![Screenshot 2023-09-13 at 16 40 08](https://github.com/o1-labs/o1js/assets/37619567/6274afa3-9d88-4438-9eb8-7b21d9cec582)

I am wondering how does webpack/angular come in play here anyway. Do you know how that file is created?

**Comment by mitschabaude** [comment]
> Do you know how that file is created?

Yes it's created on the fly from function source code, with `Function.toString()`, which is then put in a Blob, which then the worker is created from
See explanation here: https://github.com/o1-labs/o1js/pull/33

**Comment by directcuteo** [comment]
Do you know if this works with `react`? Isn't react also bundled by webpack?
